modoboa_amavis/management/commands/qcleanup.py

```diff
@@ -56,11 +56,32 @@ def handle(self, *args, **options):
             "Deleting messages older than {} days...".format(
                 conf["max_messages_age"]))
         limit = int(time.time()) - (conf["max_messages_age"] * 24 * 3600)
-        Msgs.objects.filter(time_num__lt=limit).delete()
-
+        # Delete older messages in batches.
+        # This would avoid consuming too much RAM when
+        # having to delete many thousands of messages
+        # leading to process OOM kill or soft crash.
+        res = Msgs.objects.filter(time_num__lt=limit)[:5000]
+        while res.count() != 0:
+            for item in res:
+                item.delete()
```

Comment:
Why do you delete items one by one? Have you tried with a ``res.delete()``?

Reply:
@tonioo I tried it, but got the error: `AssertionError: Cannot use 'limit' or 'offset' with delete.`

Reply:
I see. I guess something like this would be more efficient (only 2 SQL queries):
```
Msg.objects.filter(pk__in=Msg.objects.filter(time_num__lt=limit).values_list('pk', flat=True)[:5000]).delete()
```

Reply:
@tonioo Thanks, but I'm getting this error: `django.db.utils.NotSupportedError: (1235, "This version of MySQL doesn't yet support 'LIMIT & IN/ALL/ANY/SOME subquery'")`

My test server is running: `mysql  Ver 14.14 Distrib 5.7.30, for Linux (x86_64)`.

I'm not familiar with Django ORM, this is the code I arrived at that produced the above error:

```
res = Msgs.objects.filter(time_num__lt=limit)[:5000]
        while res.count() != 0:
            Msgs.objects.filter(pk__in=res.values_list("pk", flat=True)).delete() # ERROR
            res = Msgs.objects.filter(time_num__lt=limit)[:5000]
```

Any ideas? 😉

Reply:
> AssertionError: Cannot use 'limit' or 'offset' with delete.

@tonioo Do you know why Django raises this exception when `DELETE FROM <table> WHERE <condition> LIMIT <num>` is a valid Mysql query?

Reply:
No, I don't. Maybe a raw SQL query would do the trick?

Reply:
> Maybe a raw SQL query would do the trick?

Is that a good solution since we have to write two queries according to the database (Mysql / Postgres) being used?

Reply:
And what about execution time ?

Reply:
> And what about execution time ?

15-30 minutes on an instance handling 150k messages per day. Not bad since this is a daily run script, but definitely not the best. Out of that, almost 15 minutes is taken up by this single line:

```
res = Maddr.objects.annotate(
            msgs_count=Count("msgs"), msgrcpt_count=Count("msgrcpt")
        ).filter(msgs_count=0, msgrcpt_count=0)
```

... probably because table `maddr` contains `303993` records.

Reply:
Strange that it takes so much time... Can you try to catch the associated SQL query and run an explain/analyze on it?